### PR TITLE
Salesforce make `start_date` optional and change format to YYYY-MM-DD

### LINKED ISSUE #7847: Various UX improvements for Salesforce
## Tell us about the problem you're trying to solve
In the setup of the Salesforce configuration, we ask for a date as a required input. This is highly unusual as Salesforce is like a database - it should just sync everything as a full refresh initially. 

## Describe the solution you’d like
We could either remove the field altogether or make this an optional field. For Close.com, we make this an optional field. 
“start_date - The start date to sync data. Leave blank for full sync. Format: YYYY-MM-DD. (e.g. 2021-01-01) “

## Acceptance Criteria

1. Make the `start_date` optional
2. Change the pattern for start date from `YYYY-MM-DDTXX:XX:XXZ` to `YYYY-MM-DD` (but it should still accept previous format for compatbility)
2. add `title` and description to all properties in Salesforce's input spec

### PR DESCRIPTION
**Pull request by augan-rymkhan**
## What

1. Make the `start_date` optional
2. Change the pattern for start date from YYYY-MM-DDTXX:XX:XXZ to YYYY-MM-DD (but it should still accept previous format for compatbility)
3. add `title` and `description` to all properties in Salesforce's input spec

## How

1. In the current implementation `start_date` variable is used in the query and it is defined as: 
`start_date = next_page_token or stream_date or self.start_date` 
 `self.start_date` is from the user input config
  The solution provided here:  if `start_date` is not None, then add `WHERE {self.cursor_field} >= {start_date}` clause to the query, otherwise query will fetch all data without filtering by date (also there is a LIMIT clause).
2. Added `format_start_date` method will transform the `start_date` from format `2021-07-25` into the format `2021-07-25T00:00:00Z
   For backward compatibility, it works also with `start_date` which is entered in date time (YYYY-MM-DDTXX:XX:XXZ) format.
3. Added `title` for all properties in `spec.json` file.

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? If yes, please make sure to include it here and in any changelogs with the 🚨🚨 emoji
What is the end result perceived by the user?

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Review by eliziario — APPROVED** [review]

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py [hunk 1]
```diff
@@ -263,7 +263,12 @@ class IncrementalSalesforceStream(SalesforceStream, ABC):
     def __init__(self, replication_key: str, start_date: str, **kwargs):
         super().__init__(**kwargs)
         self.replication_key = replication_key
-        self.start_date = start_date
+        self.start_date = self.format_start_date(start_date)
+
+    @staticmethod
+    def format_start_date(start_date):
+        """Transform the format `2021-07-25` into the format `2021-07-25T00:00:00Z`"""
+        return pendulum.parse(start_date).strftime('%Y-%m-%dT%H:%M:%SZ')
```
Comment (vitaliizazmic) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:271]: What if start_date is None?

Comment (augan-rymkhan) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py]: @vitaliizazmic This case is fixed now.

**Review by vitaliizazmic — APPROVED** [review]

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json [hunk 2]
```diff
@@ -4,41 +4,41 @@
...
-      "client_secret",
-      "refresh_token",
-      "start_date",
-      "api_type"
-    ],
+    "required": ["client_id", "client_secret", "refresh_token", "api_type"],
     "additionalProperties": false,
     "properties": {
       "client_id": {
+        "title": "Client Id",
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json:11]:
```suggestion
        "title": "Client ID",
```

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json [hunk 3]
```diff
@@ -4,41 +4,41 @@
...
       },
       "refresh_token": {
+        "title": "Refresh Token",
         "description": "Salesforce Refresh Token used for Airbyte to access your Salesforce account. If you don't know what this is, follow this <a href=\"https://medium.com/@bpmmendis94/obtain-access-refresh-tokens-from-salesforce-rest-api-a324fe4ccd9b\">guide</a> to retrieve it.",
         "type": "string",
         "airbyte_secret": true
       },
       "start_date": {
-        "description": "UTC date and time in the format 2017-01-25T00:00:00Z. Any data before this date will not be replicated. This field uses the \"updated\" field if available, otherwise the \"created\" fields if they are available for a stream.",
+        "title": "Start Date",
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json:28]:
```suggestion
        "title": "Replication Start Date",
```

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json [hunk 4]
```diff
@@ -4,41 +4,41 @@
...
+        "description": "Date in the format 2017-01-25. Any data before this date will not be replicated. This field uses the \"updated\" field if available, otherwise the \"created\" fields if they are available for a stream.",
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json:29]: Can you also add a sentence like `If not set, then by default all your data is replicated.`

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json [hunk 5]
```diff
@@ -4,41 +4,41 @@
...
         "type": "string",
-        "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z$",
-        "examples": ["2021-07-25T00:00:00Z"]
+        "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z|[0-9]{4}-[0-9]{2}-[0-9]{2}$",
+        "examples": ["2021-07-25", "2021-07-25T00:00:00Z"]
       },
       "is_sandbox": {
+        "title": "Is Sandbox?",
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json:35]:
```suggestion
        "title": "Sandbox",
```

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json [hunk 6]
```diff
@@ -4,41 +4,41 @@
...
         "description": "Whether or not the the app is in a Salesforce sandbox. If you do not know what this, assume it is false. We provide more info on this field in the <a href=\"https://docs.airbyte.io/integrations/destinations/salesforce#is_sandbox\">docs</a>.",
         "type": "boolean",
         "default": false
       },
       "api_type": {
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json:40]: can you create an issue to deprecate this option? this is an implementation detail that we should remove eventually

**Review by sherifnada — APPROVED** [review]

**Review by sherifnada — APPROVED** [review]

**Review by keu — APPROVED** [review]

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json (hunk 6 above)
Comment (augan-rymkhan) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/spec.json:40]: It's created [here](https://github.com/airbytehq/airbyte/issues/8419)

**Pull request merged**